### DISCUSSION TITLE
How to configure inter core software interrupt priority

### DISCUSSION THREAD

**Discussion by gege0500** [polarfire-soc/polarfire-soc-documentation#230] [How to] [answered]
U54_ 1 runs  FreeRTOS system, and the other U54s run bare metal programs. If U54_ 2 and U54_ 3 call raise_soft_interrupt(1U) at the same time, U54_ 1 gives priority to processing interrupts from which hart? How to configure inter core software interrupt priority?

**Comment by hughbreslin** [comment, marked as answer]
Hi @gege0500 there is no handling provided to determine the source of a soft interrupt - the CLINT only contains a pending register to indicate a soft interrupt has been raised on a hart but doesn't indicate the hart that has raised the interrupt. So I don't think the software interrupt provides the functionality you're describing as there is only a single software interrupt input to each core and not a dedicated one for each source core. Without a register to poll and check which cores have raised interrupts that are pending I don't think this is possible. In terms of priorities the software interrupt has the highest priority, then timer, then external - the only way you can achieve individual interrupt priorities is using the PLIC which is connected to the external interrupt of the individual harts.

If I'm incorrect and it does please let me know if I've missed some information.

The way we are signaling between processors is using the IHC (Inter Hart Communication) block in the Icicle Kit Reference Design to allow us to know the source hart for a message being sent - you should check out our [AMP documentation](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/asymmetric-multiprocessing/amp.md) on this.

**Reply by gege0500** [reply]
Thank you for your reply. Is there an IHC demo for communication between U54_1 and U54_2/3?

**Reply by hughbreslin** [reply]
Yes we have an AMP demo, have a look at the "Linux + FreeRTOS/Bare Metal Configuration" section of the documentation I linked.
